### DISCUSSION TITLE
Is it possible to know how the Message Pruning of each channel is set?

### DISCUSSION THREAD

**Discussion by mpapini64** [nextgenhealthcare/connect#5389] [Q&A] [answered]
I would need to know in my channel transformer, for each deployed Channel, the value of the Message Storing Metadata if it is "Stored indefinitely" or if it is set to "Prune metadata older than: x days" and obviously the corresponding value in days.
I have searched around but there is no information ...
That you know this information is available?

**Comment by pacmano1** [comment]
```sql
select value
from configuration
where category = 'core'
  and name = 'channelMetadata'
```
You then need to parse the XML returned to find that info.   I can't recall but the rest API might include that info with a channel export.

In my opinion, I would create a code template that gets called in the global deploy script that goes and grabs the data you need from mirth's db and throws  it in the globlal map as a JSON array.   It is then accessible all the time and of course get refreshed on every channel deploy.   You can even have the array contain the exact SQL (or table names at least) needed to execute a clean up, no SP needed.   You can have a standalone channel that iterates over that array and call your clean up.   I don't think the fact an SP is compiled is all that useful for your use case of purging errors.

**Reply by mpapini64** [reply]
Thanks, I'd say it's perfect.
I know the channel ID so I pull up the query result, then look in the XML for that id and look for the values inside <pruningSettings>.
I would say it is a good solution, I will make a function that does this thing, passing the channel ID.
What I am trying to do is a channel that is called at intervals and clears the error messages in the Mirth and I wanted it to use the same settings as the Message Pruning.
Calling my SP that deletes the records in error at this point, I pass the names of the channels that have the message pruning activated with the value in days set, I was missing only this last piece!
Thanks again!

**Reply by mpapini64** [reply]
Ok, Houston we have a problem.
The value field of the query is an NVARCHAR (MAX) on MSSQL, I have never had the opportunity to work with this type of data ....
I run the query ... I check the result, but I can't read the content.
Here is the piece of code : 
```
var dbConn = dbOpenChannel ('SQLSERVER','//localhost:1433/','MirthDB3101')
if (dbConn!=null) {
	result = dbConn.executeCachedQuery("select value from configuration where category = 'core' and name = 'channelMetadata'");
	if (result.next()) 	{
		var clobField = result.getClob("value")
		var strOut = new java.lang.StringBuffer();
		strOut = clobField.getAsciiStream()
		logger.info('strOut:'+strOut);
		xmlValue = new XML (strOut)
		logger.info('xmlValue:'+xmlValue);
	}
}
```
Log response : [2022-08-27 08:43:18,369]  INFO  (transformer:?): strOut:net.sourceforge.jtds.util.BlobBuffer$AsciiInputStream@6da36df8
I'm definitely doing something wrong!

**Reply by pacmano1** [reply]
Can you CAST that field?

**Reply by mpapini64** [reply]
We also tried the cast works, but the maximum is 8000 characters and fails to pull up all the XML making it impossible to deserialize, truncating the final part of the tags.
I also tried with AS TEXT but nothing is like it was a CBLOB same previous error, while with AS VARCHAR (8000) it works, but only if you have 5 or 6 channels! :(

**Reply by pacmano1** [reply]
Dunno, run postgres and stop dealing with an expensive engine with sometimes less than stellar features.   Also not clear what driver you are using on your connection string.

**Reply by mpapini64** [reply]
Sorry what's that got to do with it?
The reading of a CBLOB field is identical whatever the db from which you read it. That field in postgres is of type TEXT so anyway I would have the same problem.
Also I have hundreds of Mirth installations it is certainly not possible to migrate to postgres, so it is easier to update them all to Mirth 4.01 which clears the error messages!

**Reply by pacmano1** [reply]
It was a joke.

**Reply by pacmano1** [reply]
Also, in postgres that column is of type text.  And no, you would not have the same problem in postgres, that was why I made a joke about it

In my engine the below returns ~260,000 characters of the 522 channels that are on the engine.

So this returns the full value as expected:

```SQL
var dbConn = GetDatabaseConnectionMirth();
var sql = "select value from configuration where category = 'core' and name = 'channelMetadata'"
var result = dbConn.executeCachedQuery(sql);
```

**Reply by pacmano1** [reply]
@ odo (not here on GH that I know so intentioanlly mistagged) over on the forums suggest https://forums.mirthproject.io/forum/mirth-connect/support/181838-javax-sql-rowset-serial-serialclob-b235c1cf-mirth-json-error?p=181849#post181849 (for Oracle at lest).  Must login to see screenshot he posted.

**Reply by mpapini64** [reply]
Thanks... eventually I found ... I made a small function:
`/**
	Letura di un campo CBLOB VARCHAR(MAX)
	
	@param 	{Recordset} 	result	- Recordset risultante di una dbConn.executeCachedQuery
	@param 	{String} 		field	- Campo db CBLOB,VARCHAR(MAX), IMAGE...
	@return 	{String} 		Contenuto del campo "field"
*/
function ReadCBlobToString (result, field) {

		var clobField = result.getClob(field)
		var strOut = new java.lang.StringBuffer();
		strOut = clobField.getAsciiStream();
		var strWriter = new StringWriter();
		IOUtils.copy(strOut, strWriter); 
		return strWriter.toString();
}

`

**Comment by odoodo** [comment, marked as answer]
> I have searched around but there is no information ...

I assume, you already managed to create an instance of the client class and to login into a server?
If so, use the following approach:
```js
// as pruning settings are part of the channel meta data, first get hold of the channel meta data
var channelMetaDataMap = client.getChannelMetadata();
// for demonstration purposes extract just 1 channel metadata (you'll have to loop for your use case)
var specificMetaData =  channelMetaDataMap.get('CHANNEL_ID');
// extract the pruning settings from the meta data
var pruningSettings = specificMetaData.getPruningSettings();

// retrieve info about when message meta data will be pruned
var metaDataPruningDelay = pruningSettings.getPruneMetaDataDays();
// retrieve info about when message content data will be pruned
var contentPruningDelay = pruningSettings.getPruneContentDays();
// retrieve info about if archiving is enabled
var archiveEnabled = pruningSettings.isArchiveEnabled();
```

If you just want to access the channel pruning settings of the server on which you are running the script, you can directly instantiate the configuration controller instead of creating a client and logging into a server:

```js
// directly accessing the local config controller. Using client as var to keep code consistent with the snippet above
var client = com.mirth.connect.server.controllers.ConfigurationController.getInstance();
```

> @ odo (not here on GH that I know so intentioanlly mistagged) 

it's odoodo at GitHub

**Reply by mpapini64** [reply]
Hi, no I've never instantiated a client class to access all channels, I've never had the opportunity.
I suppose it has a syntax similar to the one you propose to me for the channel in use?
The approach that you propose to me is very interesting as it prevents me from going to look for information from the DB.
It is clear to me how to access the individual information, but I miss instantiating the client class.

**Reply by odoodo** [reply]
If you only want to access the channels of the **same mirth instance**, you are running your script onto, the direct instantiation of the config controller will do the job as shown in my previous post:
```js
var client = com.mirth.connect.server.controllers.ConfigurationController.getInstance();
```
If you also want to access information of channels running on **other servers**, you'll need the client class. You can find an example of how to use the client class for accessing remote systems in [this post](https://forums.mirthproject.io/forum/mirth-connect/support/16298-code-template-automatic-mirth-backup).

In a nutshell, you have to do the following:
```js
try{
  // create a client instance and initialize it with the server to which it should connect
  var client = new com.mirth.connect.client.core.Client('https://' + MY-MIRTH-INSTANCE + ':8443');
  // log on to the server
  try{
    // log in with an existing user account
    var loginStatus = client.login(USERNAME, PASSWORD);
  }catch(ex){
    throw 'Unable to log-on the server  with credentials ' + username + '/' + password + ' (incompatible mirth version?): ' + ex.message;
  }
  
  // check if login was successful
  if (loginStatus.getStatus() != com.mirth.connect.model.LoginStatus.Status.SUCCESS) {
    throw 'Unable to log-on the server with credentials ' + username + '/' + password + '(status ' + loginStatus.getStatus() + ')';
  }
  
  **DO YOUR WORK HERE**

}finally{
  // end the session
  client.logout();
  // and close the client instance
  client.close();
}
```

**Reply by mpapini64** [reply]
Perfect, I did as you said and it works perfectly.
So I eliminated the DB reading to read this metadata, even though I found out how to read a CBLOB / VARCHAR (MAX) which hadn't happened to me yet.
Thank you so my error prune channel is much more elegant.

**Reply by mpapini64** [reply]
Another small change ... or rather the solution to access the metadata worked perfectly for Mirth versions higher than 3.7.1.
Unfortunately I have a handpiece of Mirth 3.4.2 for now impossible to migrate ... so I would like to do the same for that version.
But when I run the channel it tells me **getChannelMetadata** function not found.
The client is created successfully but there seems to be no access to the metadata or the function is called differently.
`var channelMetaDataMap = client.getChannelMetadata()`

**Reply by odoodo** [reply]
Unfortunately, I do not have access to the source code of Mirth 3.4.2 to check. Is it somewhere publicly available?

**Reply by mjtrac** [reply]
Try https://github.com/nextgenhealthcare/connect/tree/3.4.x

**Reply by odoodo** [reply]
The Mirth API is not consistent, which, I feel, contradicts the sense of an API. 

For older Mirth versions, this is part of ChannelProperties:

```js
// obtain channel list from server
var channels = client.getAllChannels().toArray();
// now extract the needed metadata from each channel configuration
for each (channel in channels) {
        // just in case you are wondering where to get this one from ;-)
        var currentChannelId = channel.getId(); 
	// get hold of the channel properties (which contain amongst others the prunning settings)
	var prunningSettings = channel.getProperties();

       /** from here on, it's the same code like shown before */

	// retrieve info about when message meta data will be pruned
	var metaDataPruningDelay = pruningSettings.getPruneMetaDataDays();
	// retrieve info about when message content data will be pruned
	var contentPruningDelay = pruningSettings.getPruneContentDays();
	// retrieve info about if archiving is enabled
	var archiveEnabled = pruningSettings.isArchiveEnabled();
	
	*** DO YOUR METADATA PROCESSING HERE *** 
}
```

**Comment by mjtrac** [comment]
Another approach, assuming all channels are output as one file.  You'd have to vary the amount of "after context" to the grep command, and, depending on whether this was for regular use by others, further clean it up.  For multiple files you'd need to call this against each file.  You could also use sed to update the settings in the XML files and reload them (not tested).
```
 grep -A 4 -e "channel v" -e "<pruning" .\Dummy.xml
<channel version="4.1.0">
  <id>cbd91c3b-f6c7-495f-9840-f38ba00e0325</id>
  <nextMetaDataId>2</nextMetaDataId>
  <name>Dummy</name>
  <description></description>
--
      <pruningSettings>
        <pruneMetaDataDays>1</pruneMetaDataDays>
        <archiveEnabled>true</archiveEnabled>
        <pruneErroredMessages>false</pruneErroredMessages>
      </pruningSettings>
```

**Reply by mpapini64** [reply]
The idea of exporting the channels and reading the XML was a last chance, I admit, anyway thanks for the tip!
